agent/aws_role.go

```diff
@@ -0,0 +1,144 @@
+package main
+
+import (
+//	"errors"
+	"net/url"
+	"path"
+	"sync"
+	"strings"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/hcl"
+
+	"github.com/spiffe/spire/proto/agent/nodeattestor"
+	"github.com/spiffe/spire/proto/common"
+	spi "github.com/spiffe/spire/proto/common/plugin"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/aws/aws-sdk-go/service/sts/stsiface"
+
+
+)
+
+
+const (
+	pluginName = "aws_iam_role"
+)
+
+type IamRoleConfig struct {
+	TrustDomain string `hcl:"trust_domain"`
+
+	stsClient stsiface.STSAPI
+}
+
+type IamRolePlugin struct {
+	iamRoleArn string
+	trustDomain string
+	stsClient stsiface.STSAPI
+
+	mtx *sync.RWMutex
+}
+
+func (p *IamRolePlugin) spiffeID() *url.URL {
+	spiffePath := path.Join("spire", "agent", pluginName, parseIamRole(p.iamRoleArn))
+	id := &url.URL{
+		Scheme: "spiffe",
+		Host: p.trustDomain,
+		Path: spiffePath,
+	}
+
+	return id
+}
+
+func parseIamRole (arn string) (agent_id string) {
+	tokens := strings.Split(arn, ":")
+	token_length := len(tokens) - 2
+	relevant_tokens := tokens[token_length:]
+	return strings.Join(relevant_tokens, "/")
+}
+
+func (p *IamRolePlugin) FetchAttestationData(req *nodeattestor.FetchAttestationDataRequest) (*nodeattestor.FetchAttestationDataResponse, error) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	input := sts.GetCallerIdentityInput{}
+
+	result, err := p.stsClient.GetCallerIdentity(&input)
+	if err != nil {
+		return nil, err
+	}
+
+	// pull out role arn and put it into the p.iamRoleArn
+	p.iamRoleArn = *result.Arn
```

Comment:
We take a read lock for this method, but this writes to the plugin struct.

Since this bit of information is related to a single Attestation request, and not the plugin config itself, we should probably avoid setting it on the struct. Instead, we could pass the ARN in to spiffeID as an argument